test/Microsoft.Azure.Devices.Edge.Test.Common/config/BaseModuleConfigBuilder.cs

```diff
@@ -0,0 +1,141 @@
+            {
+                this.Env[key] = value; // for duplicate keys, last save wins!
+            }
+
+            return this;
+        }
+
+        public IModuleConfigBuilder WithProxy(Option<Uri> proxy, Protocol protocol)
+        {
+            proxy.ForEach(
+                p =>
+                {
+                    string proxyProtocol;
+                    switch (protocol)
+                    {
+                        case Protocol.Amqp:
+                            proxyProtocol = Protocol.AmqpWs.ToString();
+                            break;
+                        case Protocol.Mqtt:
+                            proxyProtocol = Protocol.MqttWs.ToString();
+                            break;
+                        case Protocol.AmqpWs:
+                        case Protocol.MqttWs:
+                            proxyProtocol = protocol.ToString();
+                            break;
+                        default:
+                            throw new ArgumentException("Unknown protocol");
+                    }
+
+                    // If UpstreamProtocol was already set in this config to a non-compatible value,
+                    // throw an error. The caller will need to fix the conflict in their code.
+                    // Otherwise, replace it with the proxy-compatible value.
+                    if (this.Env.TryGetValue("UpstreamProtocol", out string u))
+                    {
+                        if (u != protocol.ToString() && u != proxyProtocol)
+                        {
+                            string message = $"Setting \"UpstreamProtocol\" to \"{proxyProtocol}\"" +
+                                             $"would overwrite incompatible value \"{u}\"";
+                            throw new ArgumentException(message);
+                        }
+                    }
+
+                    this.WithEnvironment(
+                        new[]
+                        {
+                            ("https_proxy", p.ToString()),
+                            ("UpstreamProtocol", proxyProtocol)
+                        });
+                });
+
+            return this;
+        }
+
+        public IModuleConfigBuilder WithDesiredProperties(IDictionary<string, object> properties)
+        {
+            foreach ((string key, object value) in properties)
+            {
+                this.DesiredProperties[key] = value; // for duplicate keys, last save wins!
+            }
+
+            return this;
+        }
+
+        // How is module information from this class composed into a
+        // configuration JSON document? For a module, e.g. "myModule":
+        //
+        // modulesContent
+        //   $edgeAgent
+        //     properties.desired
+        //       modules
+        //         myModule         <== this.Deployment
+        //           settings       <== this.Settings
+        //           env            <== this.Env
+        //   myModule
```

Comment:
Under modulesContent, it should contains $edgeAgent and $edgeHub only, right? myModule should exist under $edgeAgent > properties.desired > modules only, am i correct?

Reply:
No, modulesContent holds the initial desired properties for **all** module twins. Typically in our deployments we only add desired properties for agent and hub, but if your module has some settings, you could add them there. For example, see how the deployment config is built in these [edge agent tests](https://github.com/Azure/iotedge/blob/7026fd370e7d89b4754f1aa83e5b5d9358cc9de1/edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.IoTHub.Test/EdgeAgentConnectionTest.cs#L111).

Reply:
thanks for your clarification.